search-app/src/cmr/search/api/autocomplete.clj

```diff
@@ -0,0 +1,71 @@
+(ns cmr.search.api.autocomplete
+  "Defines the API for autocomplete-suggest in the CMR."
+  (:require
+    [clojure.string :as str]
+    [compojure.core :refer :all]
+    [cmr.common.log :refer (debug info warn error)]
+    [cmr.common.mime-types :as mt]
+    [cmr.common.services.errors :as svc-errors]
+    [cmr.common-app.api.routes :as common-routes]
+    [cmr.search.api.core :as core-api]
+    [cmr.search.services.autocomplete-service :as ac]
+    [cmr.search.services.query-service :as qs]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Support Functions
+
+(def page-size-default 20)
+(def page-num-default 0)
+
+(defn- string-param-or-default->int
+  [s default]
+  (if s
+    (Integer/parseInt s 10)
+    default))
+
+(defn- lower-case-and-trim
+  [s]
+  (str/lower-case (str/trim s)))
+
+(defn- get-autocomplete-suggestions
+  "Invokes Elasticsearch for autocomplete query result"
+  [ctx path-w-extension params headers]
+  (let [params (core-api/process-params :autocomplete params path-w-extension headers mt/json)
+        query (:q params)
+        type-param (:types params)]
+    (when-not query
+      (svc-errors/throw-service-errors
+       :bad-request
+       ["Missing param [q]"
+        "Usage [/autocomplete?q=<term>[&types=[&page-size=[&page-number=]]]]"
+        "q : query string to search for suggestions"
+        "types : comma separated list of types"
+        "page-size : maximum number of suggestions : default 20"
+        "page-number : results page : default 0"]))
+    (let [term (lower-case-and-trim query)
```

Comment:
The double `let`s isn't something I've seen before. Is there a reasoning behind it?

Reply:
The double let is because I need to do validation on the `query` value before running the autocomplete search. I can see if I can restructure things